pkg/resource/deploy/step_generator.go

```diff
@@ -1180,6 +1180,76 @@ func (sg *stepGenerator) getProviderResource(urn resource.URN, provider string)
 	return result
 }
 
+type replaceOnChangeSet struct {
+	m   map[resource.PropertyKey]bool
+	all bool
+}
+
+func newReplaceOnChangeSet(keys []resource.PropertyKey) *replaceOnChangeSet {
+	m := map[resource.PropertyKey]bool{}
+	all := false
+	for _, key := range keys {
+		if key == "*" {
+			all = true
+			continue
+		}
+		m[key] = true
+	}
+	return &replaceOnChangeSet{
+		m:   m,
+		all: all,
+	}
+}
+
+func (s *replaceOnChangeSet) Contains(key resource.PropertyKey) bool {
+	if s.all {
+		return true
+	}
+	if _, ok := s.m[key]; ok {
+		return true
+	}
+	return false
+}
+
+func (s *replaceOnChangeSet) All() bool {
+	return s.all
+}
+
+// calculateReplaceKeys determines the keys that would force a replacement. If there are any, a replacement
+// operation will be scheduled for the resource, else an update operation will be scheduled. The following
+// may cause a key to force a replacement:
+// * The detailed diff reports the key requires replacement
+// * The detailed diff reports the key requires update, but the key is marked as replace-on-change (or "*")
+// * The diff includes the key as a replace key
+// * The diff includes the key as a changed key, but the key is marked as replace-on-change (or "*")
+// * The old state had initErrors, and replace-on-change is "*" (a synthetic property is used to track)
+func (sg *stepGenerator) calculateReplaceKeys(
```

Comment:
I'm pretty sure this is not "right", largely because it's not clear what the exact semantic requirements are for the `plugin.DiffResult` passed back from a provider.  Is `DetailedDiff` now required and definitive?  If so, what are the required semantics of the keys in the detailed diff?  Do they always include all subkeys?  Or never?  Are the keys canonicalized in any way (`foo.bar` vs. `["foo"]["bar"]`) - or does parsing and interpretation of those key strings have to be handled by clients?  If so, what is the exact allowed syntax and semantics clients must support?

Once all that is known, I believe this should instead operate by parsing `replaceOnChanges` as a `PropertyPath` (like with `ignoreChanges`) and then applying that path to the `DetailedDiff` properties and overwriting any changes that match with replaces.

The logic here currently supports all the "normal" cases, including normal property nesting, but likely doesn't handle a variety of unusual cases and non-standard but "legal" potential implementations of `plugin.DiffResult`.

cc @pgavlin

Reply:
After talking with @pgavlin and reviewing the [Provider Implementors Guide section on detailed diffs](https://github.com/pulumi/pulumi/blob/7820d2c17b40ca2345f4215b3a6789744e2784e6/sdk/go/common/resource/plugin/provider-implementers-guide.md#detailed-diffs), I took a very different approach here.  We now interpret both the DetailedDiff keys and the `replaceOnChanges` items as paths, and then determine whether a replaceOnChange path "contains" each detailed diff key.  If it does, we change that detailed diff key to a replacement.  This is much more structurally sound, and aligned with the existing property path semantics.